File: VMSFrontend/src/app/core/document.models.ts
/** The two kinds of owner a document belongs to (FSD §23A): one engine, reached from a route of its own for each. */
export type DocumentOwnerType = 'BusinessPartner' | 'Vehicle';

/** The document type master (§23A.1): what decides everything about how one kind of document behaves. */
export interface DocumentType {
  id: number;
  code: string;
  name: string;
  appliesTo: DocumentOwnerType[];
  partnerRole?: string | null;
  isExpirable: boolean;
  defaultValidityValue?: number | null;
  defaultValidityUnit?: string | null;
  isPeriodic: boolean;
  renewalLeadDays: number;
  mandatoryLevel: 'None' | 'Warn' | 'Required';
  requiresDocumentNumber: boolean;
  hasCost: boolean;
  allowedFormats: string[];
  maxFileSizeMb: number;
  retentionYears: number;
  isActive: boolean;
  /** BR-DOC-006: the Recurring Charge Type code this type's renewal can offer to link a payment to, where one exists. */
  linkedChargeTypeCode?: string | null;
}

/** One version of a document (§23A.3), current or superseded. */
export interface DocumentVersion {
  id: number;
  documentCode: string;
  ownerType: DocumentOwnerType;
  ownerId: number;
  documentTypeId: number;
  documentTypeName?: string | null;
  versionNo: number;
  isCurrent: boolean;
  status: 'Active' | 'ExpiringSoon' | 'Expired' | 'Superseded' | 'Rejected';
  documentNumber?: string | null;
  provider?: string | null;
  issueDate?: string | null;
  expiryDate?: string | null;
  daysRemaining?: number | null;
  originalFileName: string;
  sizeBytes: number;
  rejectReason?: string | null;
  linkedTransactionId?: number | null;
  createdOn: string;
}

/** One document slot: its current version (if any) plus, on request, its history. */
export interface DocumentSlot {
  documentTypeId: number;
  documentTypeName: string;
  mandatoryLevel: 'None' | 'Warn' | 'Required';
  hasCost: boolean;
  current?: DocumentVersion | null;
  history: DocumentVersion[];
}

/** Fields sent with an upload or a renewal, as `uploadFile`'s extra form fields. */
export interface UploadDocumentFields {
  /** Upload only: a renewal's type comes from the URL, not a field. */
  documentTypeId?: string;
  documentNumber?: string;
  provider?: string;
  issueDate?: string;
  expiryDate?: string;
  /** Renew only (BR-DOC-006): the transaction already posted for this premium or fee, to tag the new version with. */
  linkedTransactionId?: string;
}

export interface DownloadLink {
  url: string;
  expiresAtUtc: string;
}

/** Filters for the Document Register (§23A.4). No branch, vehicle-category or partner-role filter yet — see the register's own note. */
export interface RegisterQuery {
  ownerType?: DocumentOwnerType | '';
  documentTypeId?: number | null;
  status?: string;
  expiringWithinDays?: number | null;
}

export interface RegisterRow {
  documentId: number;
  documentCode: string;
  ownerType: DocumentOwnerType;
  ownerId: number;
  ownerName: string;
  documentTypeId: number;
  documentTypeName: string;
  status: string;
  expiryDate?: string | null;
  daysRemaining?: number | null;
}

/** One owner missing one mandatory or periodic document (the Missing Documents report, §23A.4). */
export interface MissingDocumentRow {
  ownerType: DocumentOwnerType;
  ownerId: number;
  ownerName: string;
  documentTypeCode: string;
  documentTypeName: string;
  required: boolean;
}

/** One document due to expire within the requested month (the Expiry Calendar, §23A.4). */
export interface CalendarEntry {
  date: string;
  ownerType: DocumentOwnerType;
  ownerId: number;
  ownerName: string;
  documentTypeName: string;
  documentId: number;
}

export interface DocumentJobsResult {
  expiry: { recalculated: number; becameExpiringSoon: number; becameExpired: number };
  retention: { removed: number };
}

File: src/VMS.Modules.Documents/Models/DocumentModels.cs
namespace VMS.Modules.Documents.Models;

/// <summary>One version of a document (FSD §23A.3), current or superseded.</summary>
public class DocumentModel
{
    public int Id { get; set; }
    public string DocumentCode { get; set; } = string.Empty;
    public string OwnerType { get; set; } = string.Empty;
    public int OwnerId { get; set; }
    public int DocumentTypeId { get; set; }
    public string? DocumentTypeName { get; set; }
    public int VersionNo { get; set; }
    public bool IsCurrent { get; set; }
    public string Status { get; set; } = string.Empty;
    public string? DocumentNumber { get; set; }
    public string? Provider { get; set; }
    public DateOnly? IssueDate { get; set; }
    public DateOnly? ExpiryDate { get; set; }
    /// <summary>Days until expiry; negative once expired. Null when the type is not expirable.</summary>
    public int? DaysRemaining { get; set; }
    public string OriginalFileName { get; set; } = string.Empty;
    public long SizeBytes { get; set; }
    public string? RejectReason { get; set; }
    public int? LinkedTransactionId { get; set; }
    public DateTime CreatedOn { get; set; }
}

/// <summary>One document slot: its current version (if any) plus, on request, its history.</summary>
public class DocumentSlotModel
{
    public int DocumentTypeId { get; set; }
    public string DocumentTypeName { get; set; } = string.Empty;
    public string MandatoryLevel { get; set; } = string.Empty;
    public bool HasCost { get; set; }
    public DocumentModel? Current { get; set; }
    public List<DocumentModel> History { get; set; } = [];
}

public class UploadDocumentRequest
{
    public int? DocumentTypeId { get; set; }
    public string? DocumentNumber { get; set; }
    public string? Provider { get; set; }
    public string? IssueDate { get; set; }
    public string? ExpiryDate { get; set; }
}

/// <summary>Pre-filled from the current version by the client (BR-DOC-005); the server computes the same suggested expiry and accepts an override.</summary>
public class RenewDocumentRequest
{
    public string? DocumentNumber { get; set; }
    public string? Provider { get; set; }
    public string? IssueDate { get; set; }
    public string? ExpiryDate { get; set; }
    /// <summary>
    /// BR-DOC-006: the ledger transaction the client already posted for this premium or fee by confirming the matching recurring
    /// charge entry (the one, sole posting path — this field never posts anything itself, only tags the new version with what paid
    /// for it, for traceability). Left null when the type has no cost or nothing was linked.
    /// </summary>
    public int? LinkedTransactionId { get; set; }
}

public class RejectDocumentRequest
{
    public string? Reason { get; set; }
}

public class DownloadLinkModel
{
    public string Url { get; set; } = string.Empty;
    public DateTime ExpiresAtUtc { get; set; }
}
